src/vctrs.h

```diff
@@ -26,11 +26,61 @@ enum vctrs_type {
   vctrs_type_s3 = 255
 };
 
-enum vctrs_type vec_typeof(SEXP x);
-enum vctrs_type vec_typeof_impl(SEXP x, bool dispatch);
+/**
+ * @member type The vector type of the original data.
+ * @member proxy_method The function of the `vec_proxy()` method, if
+ *   any. This method is looked up with [vec_proxy_method()].
+ */
+struct vctrs_type_info {
+  enum vctrs_type type;
+  SEXP proxy_method;
+};
+/**
+ * @inheritMembers vctrs_type_info
+ * @member type If `proxy_method` was found, the vector type of the
+ *   proxy data. Otherwise, the vector type of the original data.
+ *   This is never `vctrs_type_s3`.
+ * @member proxy If `proxy_method` was found, the result of invoking
+ *   the method. Otherwise, the original data.
+ */
+struct vctrs_proxy_info {
```

Comment:
Should this just be`vctrs_proxy`? 

Also I'm mildly surprised it doesn't also store the restore method.

Reply:
But then, should `struct vctrs_type_info` be `struct vctrs_type`? However this would clash with `enum vctrs_type`.

About restore, that's because several operations don't need to restore, for example `vec_size()` and `vec_is_vector()`. It seems preferable to avoid retrieving too much information and spare two env lookups. If we had a generic to return all the methods of the vctrs interface in a list, I think it'd make sense to include the restore method.

Reply:
Also invoking the method in this way breaks `NextMethod()`. I don't think this is an issue for `vec_proxy()`, but it feels like `vec_restore()` methods should be able to rely on parent methods.